extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -95,13 +113,25 @@
   private final InputRowParser<ByteBuffer> parser;
   private final KafkaTuningConfig tuningConfig;
   private final KafkaIOConfig ioConfig;
+  private final Optional<ChatHandlerProvider> chatHandlerProvider;
+
+  private final Map<Integer, Long> endOffsets = new ConcurrentHashMap<>();
+  private final Map<Integer, Long> nextOffsets = new ConcurrentHashMap<>();
+  private final Lock pauseLock = new ReentrantLock();
+  private final Condition shouldPause = pauseLock.newCondition();
+  private final Condition shouldResume = pauseLock.newCondition();
 
   private volatile Appenderator appenderator = null;
   private volatile FireDepartmentMetrics fireDepartmentMetrics = null;
   private volatile boolean startedReading = false;
```

Comment:
Holy Booleans Batman!

I think it's worth looking into whether a state enum would simplify the code. I'm not sure if that will be better but I think it's worth a look.